hedera-mirror-rest-java/src/main/java/com/hedera/mirror/restjava/service/EntityServiceImpl.java

```
@@ -0,0 +1,65 @@
+
+import com.hedera.mirror.restjava.common.EntityIdParameter;
+import com.hedera.mirror.restjava.common.EntityIdType;
+import com.hedera.mirror.restjava.repository.EntityRepository;
+import jakarta.inject.Named;
+import jakarta.persistence.EntityNotFoundException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+
+@Named
+@RequiredArgsConstructor
+public class EntityServiceImpl implements EntityService {
+
+    @Value("${hedera.mirror.restJava.shard}")
+    private final long systemShard;
+
+    private final EntityRepository entityRepository;
+
+    @Override
+    public long lookup(EntityIdParameter accountId) {
+
+        long id = 0;
+
+        if ((accountId.shard() != null && accountId.shard() != systemShard) || accountId.realm() != 0) {
+            throw new IllegalArgumentException("Id %s has invalid shard or realm".formatted(accountId));
+        }
+
+        if (accountId.type() == EntityIdType.NUM && accountId.num() != null) {
+            if (!entityRepository.existsById(accountId.num().getId())) {
+                throw new EntityNotFoundException("Id %s does not exist".formatted(accountId));
+            }
+            id = accountId.num().getId();
+        } else if (accountId.type() == EntityIdType.ALIAS && accountId.alias() != null) {
+            var entityId = entityRepository.findByAlias(accountId.alias());
+            if (!entityId.isPresent()) {
+                throw new EntityNotFoundException("No account with a matching evm address found");
+            }
+            id = entityId.get();
+        } else if (accountId.type() == EntityIdType.EVMADDRESS && accountId.evmAddress() != null) {
+            var entityId = entityRepository.findByEvmAddress(accountId.evmAddress());
+            if (!entityId.isPresent()) {
+                throw new EntityNotFoundException("No account with a matching alias found");
+            }
+            id = entityId.get();
+        }
+        return id;
```

Instead of returning zero, we should throw an error if one of these three types did not validate.